framework/src/base/FEProblem.C

```diff
@@ -704,16 +709,47 @@ FEProblem::getMaxScalarOrder() const
 }
 
 void
+FEProblem::checkNonlocalCoupling()
```

Comment:
n_vars * n_subdomains * n_kernels in fact...

For Rattesnake you could have 1000 vars with 10,000 subdomains and 1000 kernels each:  That's 10 _billion_ dynamic casts to do...